spec/dynamic/layout.md

```diff
@@ -0,0 +1,335 @@
+| `struct Foo{}`[^7][^8] | `0` | `1` | Always |
+| `union Foo{f0: T0, f1: T1, ..fN; Tn}`[^6] | At least maximum of `T0`, `T1`, ..`Tn` | At least maximum of `T0`, `T1`, ..`Tn` |  |
+
+[^1]: Target defined
+
+[^2]: `usize` and pointers to `Thin` types have the same size and alignment requirement. Neither type has any padding bytes.
+
+[^3]: `fn()` refers to any fn-pointer type, regardless of return type, arguments, or ABI Tag
+
+[^4]: Range of valid values when computing as a suitably sized integer type (or pointer). Implicitly imposes a constraint of "When initialized"
+
+[^5]: References, and `Option`s of references (when `Some`) have additional aliasing constraints, see r#[dynamic.alias].
+
+[^6]: `repr(Rust)` types have an unspecified and unstable layout
+
+[^7]: Same as equivalent tuple struct. Struct size and alignment requirement may be modified by the `#[repr(align)]`
+
+[^8]: Same as equivalent unit struct
+
+## Layout Properties [§](r[dynamic.layout.properties])
+
+[§](r[dynamic.layout.properties.size]) Each `Sized` type has a property called it's size. The size is a non-negative integer.
+
+[!NOTE]: For example, the size of `u8` is `1`, and the size of `u32` is `4`.
+
+[§](r[dynamic.layout.properties.align]) Each `Sized` type and the slice type has a property called its alignment requirement.  The alignment requirement is a positive integer power of 2. An address satisfies an alignment requirement if it is a multiple of that alignment requirement. 
+
+[!NOTE]: The alignment requirement may simply be called the alignment or align of the type.
+
+[§](r[dynamic.layout.properties.size-align-value]) The size and alignment of any type is representible as a value of type `usize`. The alignment of a type divides its size.
+
+[!NOTE]: The minimum size of a type is `0`, and the minimum alignment requirement is `1`. A type that has size `0` and alignment requirement `1` is called a 1-ZST.
+
+[§](r[dynamic.layout.properties.storage]) Storage for a value is suitable for storing a value of a `Sized` type if it is at least the size of that type and the address of the start of the storage satisfies the alignment requirement.
+
+[§](r[dynamic.layout.properties.representation]) Each `Sized` type has a representation which determines how particular values of that type are layed out when read from or written to suitable storage. The representation of a type is exactly its size.
+
+[§](r[dynamic.layout.properties.padding]) The representation of a type may include padding bytes that do not participate in the computation of the value. All other bytes of the representation are called value bytes.
+
+[!NOTE]: Scalar types and pointer types have no padding bytes.
+
+[§](r[dynamic.layout.properties.validity]) Each `Sized` type has a validity invariant, which constrains the values that can be read from storage. Only value bytes are taken into account in determining the validity of a value.
```

Comment:
This is not following t-opsem terminology and also not following the structure that we've been exploring with MiniRust. I don't think the spec should be structured like this.

See [here](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#representation-relation) and [here](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#layout) for some t-opsem definitions, and [here](https://rust-lang.zulipchat.com/#narrow/stream/399173-t-spec/topic/representation.20and.20validity) for further Zulip discussion on this concern.